README.md
PiSkjerm
========

The PiSkjerm project fetches departure times from local bus stops via the Västtrafik API,
and displays the time table for upcoming departures using an Inky pHat screen, all on a Raspberry Pi Zero W.

Check out the [screen](https://shop.pimoroni.com/products/inky-phat) or the [Västtrafik API](https://developer.vasttrafik.se/) for more info.


ChangeLog.md
# Revision history for PiSkjerm

## 0.1.0.0  -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


BusTimes.hs
{-# LANGUAGE OverloadedStrings, DeriveAnyClass, DeriveGeneric,
             TypeSynonymInstances, DuplicateRecordFields #-}
module BusTimes where

import System.Environment
import Data.Aeson
import Data.Aeson.Types

import GHC.Generics (Generic)

import Data.List (nub)

import qualified Data.ByteString.Base64 as B64
import Data.ByteString.Char8 (ByteString, pack, unpack)

getAuthCredentials :: IO (String, String)
-- First we have to read the secrets from the environment,
-- so we add System.Environment to the import list and try
-- getSecret = _
-- one of he matches is the getEnv function!
-- Let's use that and get the key and secret
-- getAuthCredentials = do key <- _b "VASTTRAFIKKEY"
--                  secret <- _a "VASTTRAFIKSECRET"
--                  return (key, secret)
--
-- One of the matches is getEnv, which sounds like what
-- we want. Let's try that:
getAuthCredentials = do key <- getEnv "VASTTRAFIKKEY"
                        secret <- getEnv "VASTTRAFIKSECRET"
                        return (key, secret)
-- It works!

-- The token is the base64 encoded string "key:secret",
authToken :: String -> String -> String
authToken key secret = encodeToken $ key ++ ":" ++ secret
  where
    -- We need to base64 encode this authorization,
    -- so let's use the base64-bytesting package.
    encodeToken :: String -> String
    encodeToken token = result
      where
        -- First we need to get to bytestring
        toBS :: String -> ByteString
        -- toBs = _
        -- pack has the right type and sounds reasonable, let's use that!
        toBS = pack
        -- Next we need to turn the bytestring into base64 string
        b64ed :: ByteString
        -- b64ed = _ $ toBs token
        -- B64.encode sounds like what we need!
        b64ed = B64.encode $ toBS token
        -- And finally, we need to turn it back into a string
        result :: String
        -- result = _ b64ed
        -- unpack should to the trick!
        result = unpack b64ed

-- Now, let's try to fetch a token!

data TokenResponse = TR { scope :: String
                        , token_type :: String
                        , expires_in:: Int
                        , access_token :: String
                        } deriving (Show, Generic, FromJSON)

data BusResponse = BR { departureBoard :: DepartureBoard} deriving Show

instance FromJSON BusResponse where
  parseJSON = withObject "BusResponse" $ \v -> BR <$> v .: "DepartureBoard"

data DepartureBoard = DB { departure :: [Departure] } deriving Show

instance FromJSON DepartureBoard where
  parseJSON = withObject "DepartureBoard" $ \v -> DB <$> v .:? "Departure" .!= []


data Departure = DP { direction :: String
                    , track :: String
                    , sname :: String
                    , rtTime :: Maybe String
                    , rtDate :: Maybe String
                    , time :: String
                    , date :: String
                    } deriving (Show, Generic, FromJSON)

type Token = String
type BusStop = Integer

data BusLine = BL { name :: String
                  , track :: String
                  , departures :: [String] } deriving (Show, Eq)

getLines :: BusResponse -> [BusLine]
getLines (BR (DB dep)) = concatMap getLine names
  where names = nub $ map sname dep
        getLine name = map toLine tracks
          where dtrack = track :: Departure -> String
                tracks = nub $ map dtrack $ filter ((== name) . sname) dep
                -- filter fusion, yay!
                getTrack tr = filter ((== name) . sname)
                                . filter ((== tr) . dtrack )
                concreteTime d = case rtTime d of Just t -> t; _ -> time d
                toLine tr = BL name tr $ map concreteTime $ getTrack tr dep

-- Req networking



InkyPhat.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE CPP #-}
module InkyPhat ( runInky, InkyIO
                  -- commands
                , text, paste, display, setRotation, clear

                -- Data
                , image, size, dimensions, readString
                -- Types
                , Color (..) , Font (..), Image
    ) where 


import System.Process
import Control.Concurrent
import GHC.IO.Handle (Handle, hClose, hFlush, hFlushAll)
import Data.Maybe (fromMaybe)
import Control.Monad (join)

import Control.Monad.Reader
import Data.Text hiding (map, null)
import qualified Data.Text as T

import Prelude hiding (hPutStr, hPutStrLn, hGetLine, putStrLn,)
import Data.Text.IO

type InkyIO = ReaderT (Handle, Handle) IO

-- To avoid issues with FFI, we'll just go the easy
-- interpeter way.
sendCommand :: Text -> InkyIO ()
sendCommand cmd =
    do (stdin, _) <- ask
       dlog cmd
       lift $ do hPutStrLn stdin cmd
                 hFlushAll stdin

dlog :: Text -> InkyIO ()
# if debug
dlog = putStrLn
# else
dlog _ = return ()
#endif

readValue :: Read a => Text -> InkyIO a
readValue = fmap (read . unpack) . readString

readString :: Text -> InkyIO Text
readString cmd =
  do (stdin, stdout) <- ask
     lift $ hFlushAll stdout
     sendCommand $ "print(" <> cmd <> ")"
     val <- lift $ hGetLine stdout
     dlog val
     return val

initialCommands :: [Text]
initialCommands
    = [ "import inkyphat"
      , "from urllib import request"
      , "from datetime import datetime"]

data Color = Black | White | Red deriving (Show)

class InkyPhatVal a where
 toIPVal :: a -> Text

instance InkyPhatVal Color where
  toIPVal Black = "inkyphat.BLACK"
  toIPVal White = "inkyphat.WHITE"
  toIPVal Red   = "inkyphat.RED"

-- We only have the one font for now
data Font = Font Int

instance InkyPhatVal Font where
  toIPVal (Font a) =
    "inkyphat.ImageFont.truetype(inkyphat.fonts.PressStart2P,"
      <> (pack $ show a) <>")"

display :: InkyIO ()
display = sendCommand "inkyphat.show()"

clear :: InkyIO ()
clear = sendCommand "inkyphat.clear()"

type Image = Text

setRotation :: Int -> InkyIO ()
setRotation rot =
  sendCommand $ "inkyphat.set_rotation(" <> (pack $ show rot) <> ")"

image :: Text -> Text -> InkyIO Image
image name loc =
  do sendCommand $ name <> " = inkyphat.Image.open('" <> loc <> "')"
     return name


paste :: Image -> (Int, Int) -> InkyIO ()
paste img loc =
  sendCommand $ intercalate "," $ ["inkyphat.paste(" <> img
                                  , (pack $ show loc) <> ")"]

text :: (Int, Int) -> Text -> Maybe Color -> Maybe Font -> InkyIO ()
text xy text color font =
    sendCommand $ intercalate "," ["inkyphat.text(" <> (pack $ show xy)
                                  , "'" <> text <> "'"
                                  , toIPVal col
                                  , toIPVal fon <> ")" ]
  where fon = fromMaybe (Font 12) font
        col =  fromMaybe Black color

size :: Image -> InkyIO (Int, Int)
size img = readValue $ "(" <> img <> ".size[0], " <> img <> ".size[1])"

dimensions :: InkyIO (Int, Int)
dimensions = readValue "(inkyphat.WIDTH, inkyphat.HEIGHT)"

runInky :: InkyIO a -> IO a
runInky action =
  do (Just stdin, Just stdout , Just stderr, proc)
        <- createProcess $ cp { std_in = CreatePipe, std_out = CreatePipe
                              , std_err = CreatePipe }
     r <- flip runReaderT (stdin, stdout) $
            do mapM_ sendCommand initialCommands
               action
     mapM_ hClose [stdin, stdout, stderr]
     return r
  where cp = shell "python3 -i"


bustimes.py
from urllib import request
import json
import os
from base64 import b64encode
import logging

from datetime import datetime
lvvg = 9021014004663000
hvsnedre = 9021014003235000
authKey = os.getenv('VASTTRAFIKKEY')
authSecret = os.getenv('VASTTRAFIKSECRET')

authToken = b64encode("{k}:{s}"
                     .format(k=authKey,s=authSecret)
                     .encode('utf8')
                     ).decode('utf8')

authData = b"grant_type=client_credentials"
authUrl = "https://api.vasttrafik.se:443/token"

timeSpan = 59

logger = logging.getLogger(__name__)
logLevel = logging.ERROR
logger.setLevel(logLevel)

def getToken():
    authReq = request.Request(authUrl,
                              data=authData,
                              headers={"Authorization":
                                         "Basic {}".format(authToken)})
    resp = json.loads(request.urlopen(authReq).read().decode('utf8'))
    return (resp['access_token'])


def getInfo(stopId, track='A', token=None):
    now = datetime.now()
    date, time = tuple(now.strftime("%Y-%m-%d %H:%M").split(" "))
    params = {"stopId": stopId,
              "date": date,
              "time": time,
              "timeSpan": timeSpan}
    urlBase = "https://api.vasttrafik.se/bin/rest.exe/v2/"\
            "departureBoard?id={stopId}&date={date}&time={time}"\
            "&timeSpan={timeSpan}&maxDeparturesPerLine=2&format=json"
    url = urlBase.format(**params)
    token = getToken() if not token else token
    req = request.Request(url,
                          headers={"Authorization":
                                     "Bearer {token}".format(token=token)})
    resp = json.loads(request.urlopen(req).read().decode('utf8'))
    try:
        ds = resp['DepartureBoard']['Departure']
        deps = [ds] if type(ds) is dict else ds
        toRes = lambda dep: {'name': dep['sname'],
                             'track': dep['track'],
                             'time': dep['rtTime'], #rtTime is realtime time
                             'date': dep['rtDate'],
                             'direction':dep['direction']}
        isTrack = lambda d: d['track'] == track
        def howLong(d):
            until = (datetime.strptime(" ".join([d['date'], d['time']]),
                                       "%Y-%m-%d %H:%M") - now).seconds//60
            # Sometimes the arrival time will be in the past,
            # since the bus is almost there. Then we just show 0.
            d['until'] = until if until <= timeSpan else 0
            return d
        return list(map(howLong, map(toRes, filter(isTrack, deps))))
    except KeyError:
        # If the board is empty, we get a key error. Then we want to show
        # no times.
        return []
    except Exception as err:
        logger.error("{}:{} when handling {}".format(type(err), err, resp))
        return []


class Either(object):
    left = None
    right = None
    def __init__(self, left=None, right=None):
        self.left = left
        self.right = right
        assert(left or right)

def getBusTimes(token = None):
   logger.info("Fetching token...")
   token = getToken() if not token else token
   logger.info("Fetching Lövviksvägen...")
   lvgRes = getInfo(lvvg, token=token)
   lvgBkRes = getInfo(lvvg, track='B', token=token)
   logger.info("Fetching Hovås Nedre...")
   hvsRes = getInfo(hvsnedre, token=token)
   getTimes = lambda name, res: list(map(lambda d: d['until'],
                                         filter(lambda d: d['name'] == name,
                                                res)))
   r82 = getTimes('82', lvgRes)[:2]
   r82Bk = getTimes('82', lvgBkRes)[:2]
   rRosa = getTimes('ROSA', hvsRes)[:2]
   r158 = getTimes('158', hvsRes)[:2]
   if r158:
     rLast = Either(left=r158)
   else:
     rLast = Either(right=r82Bk)
   return [r82, rRosa, rLast]


if __name__ == "__main__":
    logger.info(getBusTimes())


piskjerm.py
#!/usr/bin/env python3

import inkyphat
import bustimes
import time
from datetime import datetime, timedelta
import logging
from bustimes import Either

from urllib.error import HTTPError

fontheight = 12
fontwidth = fontheight
linespace = 2

# Set the default rotation
inkyphat.set_rotation(180)
vt = inkyphat.Image.open("vt.png")

font = inkyphat.ImageFont.truetype(inkyphat.fonts.PressStart2P, fontwidth)

bigfont = inkyphat.ImageFont.truetype(inkyphat.fonts.PressStart2P, 14)
logger = logging.getLogger(__name__)

logLevel = logging.ERROR
logger.setLevel(logLevel)

def printlines(msg, pos = None, color=inkyphat.BLACK):
  lines = msg.splitlines()
  if pos is not None:
    x,y = pos
  else:
    x,y = (2, inkyphat.HEIGHT//2 - fontheight*len(lines)//2)
  for i, l in enumerate(lines):
    inkyphat.text((x,y+ i*(fontheight + linespace)),l,color, font)

def printBusTimes(times):
   prefix = ["82:","Rosa:", bustimes.Either("158:","82B:")]
   m = "min"
   op = []
   toL5 = lambda s: " "*(max(5-len(s),0)) + s
   for i,ts in enumerate(times):
     if type(ts) is bustimes.Either:
       pf = prefix[i].left if ts.left else prefix[i].right
       next_times = ts.left if ts.left else ts.right
     else:
       pf = prefix[i]
       next_times = ts
     s = [pf, "-","-"]
     for j,t in enumerate(next_times):
       s[j+1] = "Núna!" if t == 0 else str(t) + m
     op.append(" ".join(map(toL5, s)))
   bt = "\n".join(op)
   inkyphat.clear()
   td = datetime.now().strftime("%Y-%m-%d %H:%M")
   inkyphat.text(((inkyphat.WIDTH-16*fontwidth)//2, 18), td, inkyphat.RED, font)
   inkyphat.text((inkyphat.WIDTH//2 - 14*fontwidth//2,2)
                , "Strætóferðir", inkyphat.RED, bigfont)
   printlines(bt)
   vtc = ((inkyphat.WIDTH - 10*14 -vt.size[0] - 10)//2, inkyphat.HEIGHT-2-14)
   inkyphat.text( (vtc[0]+vt.size[0]+2,vtc[1]-4)
                , "Västtrafik", inkyphat.RED, bigfont)
   inkyphat.paste(vt,((vtc[0],vtc[1]-vt.size[1]+14-2)))
   inkyphat.show()


if __name__ == "__main__":
  frequency = timedelta(0,60)
  token = None
  while True:
    now = datetime.now()
    logger.info("Waking up at {now}".format(now=now.strftime("%Y-%m-%d %H:%M")))
    logger.info("Fetching bus departures...")
    token = bustimes.getToken() if not token else token
    try:
      bt = bustimes.getBusTimes(token)
    except HTTPError as err:
      # If the code is 401 forbidden, we need to refresh our token
      if err.code == 401:
        token = bustimes.getToken()
        bt = bustimes.getBusTimes(token)
      else:
        bt = [[],[],[]]
        logger.error("{}:{}".format(type(err),err))
    logger.info("Bus times are: {}".format(bt))
    logger.info("Outputting bus departures...")
    printBusTimes(bt)
    now_again = datetime.now()
    seconds_until = max((now+frequency - now_again).seconds, 0)
    logger.info("Sleeping for {}secs...".format(seconds_until))
    time.sleep(seconds_until)


PythonHacks.hs
{-# LANGUAGE OverloadedStrings  #-}

module PythonHacks (
 urlRequest, pyGetToken, pyGetBusTimes, pyGetDateTime
) where

import Data.ByteString (ByteString)
import qualified Data.ByteString as B
import qualified Data.Text.Encoding as E

import qualified Data.Text as T
import Data.Text hiding (map, null)

import Prelude hiding (hPutStr, hPutStrLn, hGetLine, putStrLn)
import qualified Data.Aeson as JSON

import Control.Monad.Reader

import BusTimes
import InkyPhat

-- The Nice request libraries segfault on the pi,
-- so we'll just use the already available python interpreter
data AuthHeader = Basic String | Bearer String

instance Show AuthHeader where
  show (Basic s) = show ("Basic " <> s)
  show (Bearer s) = show ("Bearer " <> s)


urlRequest :: Text -> [(Text,Text)] ->
              AuthHeader -> ByteString -> InkyIO Text
urlRequest url params auth body
 = readString readCmd
 where reqCmd = "request.Request(" <> intercalate "," args <> ")"
       readCmd = "request.urlopen(" <> reqCmd <> ").read().decode('utf8')"
       args = [ (pack $ show urlArg)
              , "headers={'Authorization':" <> (pack $ show auth) <> "}"]
              <> (if B.null body then mempty else ["data=b'" <> E.decodeUtf8 body <> "'"])
       urlArg = if null params then url
                else (url <> "?"<> ( intercalate "&" $ map paramToUrl params))
       paramToUrl (name,arg) = name <> "=" <> arg


-- | We use these to get the information via Python,
-- since the network libraries segfault on the pi
-- unless the TZ variable is set to a non-file base timezone.
-- Use these if you must.
pyGetToken :: InkyIO (Maybe TokenResponse)
pyGetToken = do (key, secret) <- lift getAuthCredentials
                let auth = authToken key secret
                    url = "https://api.vasttrafik.se/token"
                    body = "grant_type=client_credentials"
                JSON.decodeStrict . E.encodeUtf8 <$> urlRequest url [] (Basic auth) body

pyGetBusTimes :: Token -> BusStop -> InkyIO (Maybe BusResponse)
pyGetBusTimes token stop =
   do (date, time) <- pyGetDateTime
      res <- E.encodeUtf8 <$> urlRequest url (params date time) (Bearer token)  ""
      case JSON.eitherDecodeStrict res of
        Left err -> lift $ print err >> return Nothing
        Right v -> return v
  where url = "https://api.vasttrafik.se/bin/rest.exe/v2/departureBoard"
        auth = Bearer token
        params date time = [ ("id", pack $ show stop)
                           , ("maxDeparturesPerLine", pack $ show 2)
                           , ("format","json")
                           , ("timeSpan", pack $ show 59)
                           , ("date", date)
                           , ("time", time)]


pyGetDateTime :: InkyIO (Text, Text)
pyGetDateTime = do d <- readString "datetime.now().strftime('%Y-%m-%d')"
                   t <- readString "datetime.now().strftime('%H:%M')"
                   return (d,t)


ReqNetworking.hs
{-# LANGUAGE OverloadedStrings #-}
module ReqNetworking (
  hsGetToken, hsGetDateTime, hsGetBusTimes
) where

import BusTimes

-- Req networking
import Network.HTTP.Req
import Data.Default
import Control.Monad.IO.Class
import Data.Time
import Data.Time.Format.ISO8601
import Data.Text (Text)
import qualified Data.Text as T
import Control.Retry

import qualified Data.ByteString.Char8 as B
import qualified Data.Aeson as JSON

config :: HttpConfig
config = def {httpConfigRetryPolicy = delay }
  where
    -- We default to delaying for 5 sec and trying 10 times.
    delay :: RetryPolicy
    delay = constantDelay (5000*1000) <> limitRetries 5
hsGetToken :: IO (Maybe TokenResponse)
hsGetToken = fmap Just $ runReq config $
  do (key, secret) <- liftIO getAuthCredentials
     let auth = basicAuth (B.pack key) (B.pack secret)
         url = https "api.vasttrafik.se" /: "token"
         -- We need a body, but how? Let's ask!
         body :: ReqBodyBs
         -- body = _ ("grant_type=client_credentials" :: ByteString)
         body = ReqBodyBs "grant_type=client_credentials"
     responseBody <$> req POST url body jsonResponse auth

hsGetDateTime :: IO (Text, Text)
hsGetDateTime =
 do now <- zonedTimeToLocalTime <$> getZonedTime
    return (toDate now, toTime now)
  where toDate = T.pack . iso8601Show . localDay
        toTime = T.pack . Prelude.take 5 . iso8601Show . localTimeOfDay

hsGetBusTimes :: Token -> BusStop -> IO (Maybe BusResponse)
hsGetBusTimes token stop = runReq config $
  do (date, time) <- liftIO hsGetDateTime
     res <- responseBody <$> req GET url NoReqBody bsResponse (auth <> params date time)
     case JSON.eitherDecodeStrict res of
        Left err -> liftIO $ print err >> return Nothing
        Right v -> return v
  where url = https "api.vasttrafik.se" /: "bin" /: "rest.exe" /: "v2" /: "departureBoard"
        auth = oAuth2Bearer $ B.pack token
        params date time = "id" =: stop
                        <> "maxDeparturesPerLine" =: (2 :: Integer)
                        <> "format" =: ("json" :: String)
                        <> "timeSpan" =: (59 :: Integer)
                        <>  "date" =: date
                        <>  "time" =: time


Main.hs
{-# LANGUAGE TypeApplications, OverloadedStrings, CPP #-}
module Main where

import BusTimes
import InkyPhat

import Data.Maybe
import Control.Monad.Reader (lift)
import Data.Aeson
import Data.Time
import qualified Data.ByteString.Lazy.Char8 as B (pack)
import Data.Text hiding ( map, concatMap, zip
                        , length, replicate, any
                        , filter, null, take)
import qualified Data.Text as T
import Control.Concurrent
import Data.Text.Encoding as E
import System.Environment

import Data.List (sort)


-- We have a flag in the cabal file that allows us to chose between using the
-- python interpreter for time and network queries or using a native haskell
-- implementation. However, the native implementation segfaults when run on
-- the device, due to a bug in the cross-compilation toolchain, hence the
-- possiblity to choose.
# if PYTHON_HACKS
import PythonHacks
# else
import ReqNetworking
# endif

getToken :: InkyIO (Maybe TokenResponse)
getDateTime :: InkyIO (Text, Text)
getBusTimes :: Token -> BusStop -> InkyIO (Maybe BusResponse)
# if PYTHON_HACKS
getToken = pyGetToken
getDateTime = pyGetDateTime
getBusTimes = pyGetBusTimes
#else
getToken = lift hsGetToken
getDateTime = lift hsGetDateTime
getBusTimes t s = lift (hsGetBusTimes t s)
#endif


löviksVägen :: BusStop
löviksVägen = 9021014004663000
hovåsNedre :: BusStop
hovåsNedre = 9021014003235000

nearby :: [BusStop]
nearby = [löviksVägen, hovåsNedre]

setup :: InkyIO Image
setup = setRotation 180 >> image "vt" "vt.png"

fontheight :: Int
fontheight = 12
fontwidth = fontheight
linespace :: Int
linespace = 2


-- From Gabriel Gonzalez's tweet.
leftpad :: Char -> Text -> Int -> Text
leftpad c i n = T.replicate (n - T.length i) (singleton c) <> i

rightpad :: a -> [a] -> Int -> [a]
rightpad c i n = i ++ replicate (n - length i) c 

printBusTimes :: [Text] -> InkyIO ()
printBusTimes msgs =
  do (inkyW, inkyH) <- dimensions
     let (x,y) = (2, inkyH `div` 2 - fontheight * length msgs `div` 2)
         pr (m,i) = text pos m Nothing Nothing  
          where pos = (x, y + i*(fontheight + linespace))
     mapM_ pr $ zip msgs [0.. ]
  where pos = fromMaybe

printTime :: InkyIO ()
printTime = do (date, time) <- getDateTime
               (inkyW, _) <- dimensions
               let pos = ((inkyW - 16*fontwidth) `div` 2,18)
               text pos (T.unwords [date, time]) (Just Red) Nothing

printName :: InkyIO ()
printName = do (inkyW, _) <- dimensions
               let pos = ((inkyW - 14*fontwidth) `div` 2,2)
               -- Strætóferðir means "Bus Trips"
               text pos "Strætóferðir" (Just Red) (Just (Font 14))

printLogo :: Image -> InkyIO ()
printLogo img
 = do (inkyW, inkyH) <- dimensions
      (logoW, logoH) <- size img
      let (vx,vy) =  ((inkyW - 10*14 - logoW -10) `div` 2, inkyH -2-14)
          posText = (vx + logoW+2, vy -4)
          posLogo = (vx, vy-logoH+14-2)
      -- Västtrafik is the name of the Gothenburg transport authority.
      text posText "Västtrafik" (Just Red) (Just (Font 14))
      paste img posLogo

updateDisplay :: Image -> [Text] -> InkyIO ()
updateDisplay img times
 = do { clear
      ; printName
      ; printTime
      ; printBusTimes times
      ; printLogo img
      ; display }

-- Note [Known Lines]
-- We only want to show known lines, i.e. those that we take from our house,
-- and we want to display 82 first, then Rosa, and then the others.
-- Additionally, we want to show the 158 when it is running (in the morning),
-- but otherwise show the 82 going to Brottkärr, since we can only fit 3
-- lines comfortably on the display at once. This is why we do `take 3`
-- when choosing what to output, which selects the 158 if availble, but
-- the 82B otherwise.
data KnownLine = B82 [String]
               | BRosa [String]
               | B158 [String]
               | B82B [String] deriving (Eq, Ord)

-- These are the lines going in the right direction that we are interested in.
toKnownLine :: BusLine -> Maybe KnownLine
toKnownLine (BL "82" "A" d) = Just $ B82 d
toKnownLine (BL "82" "B" d) = Just $ B82B d
toKnownLine (BL "ROSA" "A" d) = Just $ BRosa d
toKnownLine (BL "158" "A" d) = Just $ B158 d
toKnownLine (BL n t d) = Nothing

knownDepartures :: KnownLine -> [String]
knownDepartures (B82 d) = d
knownDepartures (B82B d) = d
knownDepartures (BRosa d) = d
knownDepartures (B158 d) = d

instance Show KnownLine where
  show (B82 _) = "82"
  show (B82B _) = "82B"
  show (BRosa _) = "ROSA"
  show (B158 _) = "158"

renderLine :: TimeOfDay -> KnownLine -> Text
renderLine now kl =
   T.unwords $ map to5 $ [pack (show kl) <> ":"] <> dpts
 where to5 s = leftpad ' ' s 5
       dpts = rightpad "-" (map toDp $ knownDepartures kl) 2
       -- We want to show how many minutes there are until the bus departs.
       -- Note that we check for > 0, since the date might be in the past
       -- if the bus is waiting at the stop or too much time has passed
       -- since the request was made.
       toDp t = if mins > 0 then pack (show mins) <> "min"
                else "Núna!" -- "Núna" means "Right now".
         where mins = floor (lt - nowSecs) `div` 60
               lt = timeOfDayToTime $ read @TimeOfDay (t <> ":00")
               nowSecs = timeOfDayToTime now

loop :: Image -> InkyIO ()
loop img =
  do (date, time) <- getDateTime
     -- We could save the token and re-use it if it hasn't expired, but since
     -- the performance is limited by the screen refresh time anyway, we don't
     -- bother.
     mb_token <- fmap access_token <$> getToken
     case mb_token of
       Just token ->
         do stops <- catMaybes <$> mapM (getBusTimes token) nearby
            let now = read @TimeOfDay ( unpack time ++ ":00")
                -- See Note [Known Lines]
                toDisplay = take 3 $ sort $ mapMaybe toKnownLine
                              $ concatMap getLines stops
                busTimes = map (renderLine now) toDisplay
                -- "Engar ferðir núna" means "No trips right now", which we
                -- display when the buses have stopped running (usually quite
                -- late at night).
                msg = if null busTimes then ["Engar ferðir núna!"] else busTimes
            updateDisplay img msg
       _ -> return ()
     lift $ threadDelay $ 60 * 1000 * 1000
     loop img

main :: IO ()
main = runInky $ setup >>= loop
